pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/funnel/FunnelMaxStepAggregationFunction.java

```diff
@@ -0,0 +1,302 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.query.aggregation.function.funnel;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.pinot.common.request.context.ExpressionContext;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.core.common.BlockValSet;
+import org.apache.pinot.core.query.aggregation.AggregationResultHolder;
+import org.apache.pinot.core.query.aggregation.ObjectAggregationResultHolder;
+import org.apache.pinot.core.query.aggregation.function.AggregationFunction;
+import org.apache.pinot.core.query.aggregation.groupby.GroupByResultHolder;
+import org.apache.pinot.core.query.aggregation.groupby.ObjectGroupByResultHolder;
+import org.apache.pinot.segment.spi.AggregationFunctionType;
+
+
+public class FunnelMaxStepAggregationFunction
+    implements AggregationFunction<PriorityQueue<FunnelStepEvent>, Long> {
+  private final ExpressionContext _timestampExpression;
+  private final long _windowSize;
+  private final List<ExpressionContext> _stepExpressions;
+  private final Set<Mode> _modes = new HashSet<>();
+  private final int _numSteps;
+
+  public FunnelMaxStepAggregationFunction(List<ExpressionContext> arguments) {
+    int numArguments = arguments.size();
+    Preconditions.checkArgument(numArguments > 2,
+        "FUNNELMAXSTEP expects >= 3 arguments, got: %s. The function can be used as "
+            + "funnelMaxStep(timestampExpression, windowSize, ARRAY[stepExpression, ..], [mode, [mode, ... ]])",
+        numArguments);
+    _timestampExpression = arguments.get(0);
+    _windowSize = arguments.get(1).getLiteral().getLongValue();
+    Preconditions.checkArgument(_windowSize > 0, "Window size must be > 0");
+    ExpressionContext stepExpressionContext = arguments.get(2);
+    if (stepExpressionContext.getFunction() != null) {
+      // LEAF stage init this function like funnelmaxstep($1,'1000',arrayValueConstructor($2,$3,$4,...))
+      _stepExpressions = stepExpressionContext.getFunction().getArguments();
+    } else {
+      // Intermediate Stage init this function like funnelmaxstep($1,'1000',__PLACEHOLDER__)
```

Comment:
Is this a bug? Why do we get different expression in intermediate stage?

Reply:
The issue is that `stepExpressionContext` is modeled as an array function of multiple expressions.
But the intermediate stage fills it with a literal `__PLACEHOLDER__`, not a function.

Reply:
Understood. What I'm asking is that should we fix this behavior? I'd imagine currently we handle it this way as a workaround

Reply:
Correct, before this, the intermediate stage expects all literals afterward.
I don't see this could be an easy fix in this PR.

Reply:
I will create another PR after this.
This is actually a blocker for my next PR.